security/pkg/k8s/chiron/utils.go

```diff
@@ -185,321 +151,142 @@ func isTCPReachable(host string, port int) bool {
-	usages []certv1.KeyUsage, numRetries int, requestedLifetime time.Duration) (string, *certv1.CertificateSigningRequest,
-	*certv1beta1.CertificateSigningRequest, error,
-) {
-	var lastErr error
-	useV1 := true
-	csrName := ""
-	for i := 0; i < numRetries; i++ {
-		if csrName == "" {
-			csrName = GenCsrName()
-		}
-		if useV1 && len(usages) > 0 && len(signerName) > 0 && signerName != "kubernetes.io/legacy-unknown" {
-			log.Debugf("trial %v using v1 api to create CSR (%v)", i+1, csrName)
-			csr := &certv1.CertificateSigningRequest{
-				// Username, UID, Groups will be injected by API server.
-				TypeMeta: metav1.TypeMeta{Kind: "CertificateSigningRequest"},
-				ObjectMeta: metav1.ObjectMeta{
-					Name: csrName,
-				},
-				Spec: certv1.CertificateSigningRequestSpec{
-					Request:    csrData,
-					Usages:     usages,
-					SignerName: signerName,
-				},
-			}
-			if requestedLifetime != time.Duration(0) {
-				csr.ObjectMeta.Annotations = map[string]string{RequestLifeTimeAnnotationForCertManager: requestedLifetime.String()}
-			}
-			v1req, err := clientset.CertificatesV1().CertificateSigningRequests().Create(context.TODO(), csr, metav1.CreateOptions{})
-			if err == nil {
-				return csrName, v1req, nil, nil
-			}
-			lastErr = err
-			if kerrors.IsAlreadyExists(err) {
-				csrName = ""
-				continue
-			} else if kerrors.IsNotFound(err) {
-				// don't attempt to use older api unless we get an API error
-				useV1 = false
-			} else {
-				continue
-			}
-		}
-		// Only exercise v1beta1 logic if v1 api was not found
-		log.Debugf("trial %v using v1beta1 api for csr %v", i+1, csrName)
-		// convert relevant bits to v1beta1
-		v1beta1csr := &certv1beta1.CertificateSigningRequest{
-			ObjectMeta: metav1.ObjectMeta{
-				Name: csrName,
-			},
-			Spec: certv1beta1.CertificateSigningRequestSpec{
-				SignerName: &signerName,
-				Request:    csrData,
-			},
-		}
-		for _, usage := range usages {
-			v1beta1csr.Spec.Usages = append(v1beta1csr.Spec.Usages, certv1beta1.KeyUsage(usage))
-		}
-		if requestedLifetime != time.Duration(0) {
-			v1beta1csr.ObjectMeta.Annotations = map[string]string{RequestLifeTimeAnnotationForCertManager: requestedLifetime.String()}
-		}
-		// create v1beta1 certificate request
-		v1beta1req, err := clientset.CertificatesV1beta1().CertificateSigningRequests().Create(context.TODO(), v1beta1csr, metav1.CreateOptions{})
-		if err == nil {
-			return csrName, nil, v1beta1req, nil
-		}
-		lastErr = err
-		if kerrors.IsAlreadyExists(err) {
-			csrName = ""
-		}
+func submitCSR(client clientset.Interface, csrData []byte, signerName string, usages []cert.KeyUsage, requestedLifetime time.Duration) (*cert.CertificateSigningRequest, error) {
+	log.Debugf("create CSR for signer %v", signerName)
+	csr := &cert.CertificateSigningRequest{
+		// Username, UID, Groups will be injected by API server.
+		TypeMeta: metav1.TypeMeta{Kind: "CertificateSigningRequest"},
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: "csr-workload-",
```

Comment:
use GenerateName, then api server ensures there is no name conflict, so we do not need a retry